services/QuillLMS/app/services/analytics/cache_segment_school_data.rb

```diff
@@ -0,0 +1,97 @@
+    ACTIVE_TEACHERS_AT_SCHOOL_THIS_YEAR = 'ACTIVE_TEACHERS_AT_SCHOOL_THIS_YEAR',
+    ACTIVE_STUDENTS_AT_SCHOOL_THIS_YEAR = 'ACTIVE_STUDENTS_AT_SCHOOL_THIS_YEAR',
+    TOTAL_ACTIVITIES_COMPLETED_BY_STUDENTS_AT_SCHOOL_THIS_YEAR = 'TOTAL_ACTIVITIES_COMPLETED_BY_STUDENTS_AT_SCHOOL_THIS_YEAR'
+  ]
+
+  CACHE_LIFE = 60*60*25
+
+  def initialize(school)
+    @school = school
+  end
+
+  def read(namespace)
+    Rails.cache.read(@school.id, namespace: namespace)
+  end
+
+  def write(data, namespace)
+    Rails.cache.write(@school.id, data, namespace: namespace, expires_in: CACHE_LIFE)
+  end
+
+  def set_all_fields
+    set_total_teachers_at_school
+    set_total_students_at_school
+    set_total_activities_completed_by_students_at_school
+    set_active_teachers_at_school_this_year
+    set_active_students_at_school_this_year
+    set_total_activities_completed_by_students_at_school_this_year
+  end
+
+  def set_total_teachers_at_school
+    write(teachers_at_school.count, TOTAL_TEACHERS_AT_SCHOOL)
+  end
+
+  def set_total_students_at_school
+    write(students_at_school.count, TOTAL_STUDENTS_AT_SCHOOL)
+  end
+
+  def set_total_activities_completed_by_students_at_school
+    write(activities_completed_by_students_at_school.count, TOTAL_ACTIVITIES_COMPLETED_BY_STUDENTS_AT_SCHOOL)
+  end
+
+  def set_active_teachers_at_school_this_year
+    write(active_teachers_at_school_this_year.count, ACTIVE_TEACHERS_AT_SCHOOL_THIS_YEAR)
+  end
+
+  def set_active_students_at_school_this_year
+    write(active_students_at_school_this_year.count, ACTIVE_STUDENTS_AT_SCHOOL_THIS_YEAR)
+  end
+
+  def set_total_activities_completed_by_students_at_school_this_year
+    write(activities_completed_by_students_at_school_this_year.count, TOTAL_ACTIVITIES_COMPLETED_BY_STUDENTS_AT_SCHOOL_THIS_YEAR)
+  end
+
+  def active_teachers_at_school_this_year
+    teachers_at_school
+      .where("last_sign_in > ?", School.school_year_start(Time.current))
+  end
+
+  def active_students_at_school_this_year
+    students_at_school
+      .where("last_sign_in > ?", School.school_year_start(Time.current))
+  end
+
+  def activities_completed_by_students_at_school_this_year
+    activities_completed_by_students_at_school
+      .where("completed_at > ?", School.school_year_start(Time.current))
+  end
+
+  def activities_completed_by_students_at_school
+    @activities_completed_by_students_at_school ||= ActivitySession
+      .completed
+      .where(user_id: students_at_school.ids)
+  end
+
+  def students_at_school
+    @students_at_school ||= User
+      .joins(:students_classrooms)
+      .where(:students_classrooms => { classroom_id: classrooms_at_school.ids })
+  end
+
+  def classrooms_at_school
+    @classrooms_at_school ||= Classroom
+      .joins(:classrooms_teachers)
+      .where(:classrooms_teachers => { user_id: teachers_at_school.ids })
+  end
+
+  def teachers_at_school
+    @teachers_at_school ||= @school.users
```

Comment:
There may be an n+1 issue here, since you don't currently use `includes` on line 16. Would you mind verifying? Thanks

Reply:
I'm not sure how to verify, could you provide guidance? Line 16 isn't a lookup, it's just a definition.

Reply:
One of the easiest ways is just to run a sample execution in rails console, which, in development mode, prints the SQL statements: 
<img width="1030" alt="Screenshot 2023-03-21 at 1 28 19 PM" src="https://user-images.githubusercontent.com/90669/226720364-444b070e-b41d-4bf3-b44f-97f8c3e7fb99.png">

Notice that `d.teachers_at_school` triggers zero queries, whereas if you don't use `includes`, it will trigger a query.

In this case, there isn't an n+1 issue. However, I still think it's best to pre load the `users` relationships before passing the `school` to `CacheSegmentSchoolData`, since in the future we may implement a method like `admins_at_school` which could call `school.users` under the hood. The end result would be that the `school.users` query is called at least twice.